### PR TITLE
Add support for lists in kwargs (in operations)

### PR DESCRIPTION
**Pull request by thisac**
**Context:** Currently Blackbird doesn't support lists in keyword arguments in operations which causes problems with e.g. the `select` and `dark_counts`; for example, `MeasureHomodyne(select=[0, 1])` or `MeasureFock(dark_counts=[1, 0.5, 3])`.

**Description:** A new parser rule called `vallist` is added, and the `kwarg` parser rule is updated to support it. `_get_arguments` in `auxiliary.py` is also updated to support the handling of lists as keyword arguments.

**Benefits:** Lists can now be used in operations as values for keyword arguments. This allows e.g. the loading and parsing of Blackbird scripts containing this. 

**Drawbacks:** N/A

### CODE REVIEW

File: blackbird_python/blackbird/blackbirdLexer.py [hunk 1]
```diff
@@ -1,10 +1,11 @@
-# Generated from blackbird.g4 by ANTLR 4.7.1
+# Generated from src/blackbird.g4 by ANTLR 4.8
```
Comment (josh146) [inline blackbird_python/blackbird/blackbirdLexer.py:1]: Out of curiosity, does this affect the version of `antlr4-python3-runtime` we are currently pinning in setup.py and requirements.txt? We're currently pinning `4.7.1`, but this might require us to upgrade the requirements

File: blackbird_python/blackbird/tests/test_listener.py [hunk 2]
```diff
@@ -256,6 +256,13 @@ def test_operation_kwarg(self, parse_input_mocked_metadata):
             {"modes": [0], "op": "Coherent", "args": [], "kwargs": {"alpha": -0.3 + 2j}}
         ]
 
+    def test_operation_kwarglist(self, parse_input_mocked_metadata):
+        """Test that an operation with keyword arguments is correctly parsed"""
+        bb = parse_input_mocked_metadata("MeasureFock(dark_counts=[1, 3]) | [0, 1]\n")
+        assert bb.operations == [
+            {"modes": [0, 1], "op": "MeasureFock", "args": [], "kwargs": {"dark_counts": [1, 3]}}
+        ]
```
Comment (josh146) [inline blackbird_python/blackbird/tests/test_listener.py:264]: 🔥

File: requirements.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
-antlr4-python3-runtime==4.7.2
+antlr4-python3-runtime==4.8
```
Comment (josh146) [inline requirements.txt:1]: Answered my question from above 🙂

File: blackbird_python/blackbird/blackbirdLexer.py (hunk 1 above)
Comment (thisac) [reply blackbird_python/blackbird/blackbirdLexer.py:1]: I updated this in the requirements since I happened to use `antlr4-python3-runtime=4.8` while generating the files. Could there be any other issues with this that I'm missing?

Comment (josh146) [reply blackbird_python/blackbird/blackbirdLexer.py:1]: Nope, should be fine! I shouldn't have commented until I read the PR to the end 😆

**Comment by thisac** [comment]
> Your solution looks clear and concise as well  This doesn't allow lists for args, but I don't think that's too much of an issue (I don't think there is any case where an non-named argument will be passed as a list).

Thanks @josh146. I deliberately chose not to implement list support for args, since I'm not sure if that's currently needed or wanted. I chose to focus on the specific problem with the `select` and `dark_counts` kwargs. It should be very quick to implement in the future though if needed.

> Couple of requests, however:
> 
> * Don't forget to update `setup.py` with the new `antlr4-python3-runtime` version!
> * Might be worth also testing the case `dark_counts=[1]`, to ensure that it works correctly for a list of length 1
> * In addition, it would be good to test an operation that contains a combination of args, kwargs, and kwargs with lists. If none exists, I think it's fine just to make one up for the test.

Done! I expanded the tests to include testing single element lists and empty lists, as well as combinations of args, kwargs and kwargs with lists.

**Comment by josh146** [comment]
@thisac: I had a look at the coverage report. My initial instinct was that Codecov was messing up, but it looks like these two lines are not covered:
```python
elif v.nonnumeric():
    kwargslist.append(_literal(v.nonnumeric()))
```

It should be sufficient to simply add lists that contain bools or strings to your parametrized `dc` data. Once done, happy for this to be merged in!

**Comment by thisac** [comment]
Thanks @josh146 for finding the testing issue. I looked at it quickly before and also thought it was Codecov doing weird things again. Added a small variety of cases with strings and bools in the lists. Seems to pass now! :smile:
